File: src/mcp_secure_remote/url_security.py
"""URL validation helpers to block SSRF to private/restricted hosts."""
from __future__ import annotations

import ipaddress
from dataclasses import dataclass
from urllib.parse import urlparse


@dataclass(frozen=True)
class UrlSecurityOptions:
    allow_private_urls: bool = False


def validate_remote_url(server_url: str, options: UrlSecurityOptions) -> None:
    """Reject private, link-local, and metadata hosts unless explicitly allowed."""
    if options.allow_private_urls:
        return

    parsed = urlparse(server_url)
    host = _normalize_hostname(parsed.hostname or "")
    if _is_restricted_host(host):
        raise ValueError(
            f'Refusing to connect to private or restricted host "{host}"; '
            "pass --allow-private-urls for local/dev targets."
        )


def _normalize_hostname(hostname: str) -> str:
    trimmed = hostname.strip().lower()
    if trimmed.startswith("[") and trimmed.endswith("]"):
        return trimmed[1:-1]
    return trimmed


def _is_restricted_host(host: str) -> bool:
    if not host:
        return False
    if host == "localhost" or host.endswith(".localhost"):
        return True
    if host in {"0.0.0.0", "::"}:
        return True

    try:
        addr = ipaddress.ip_address(host)
    except ValueError:
        return False

    return bool(
        addr.is_private
        or addr.is_loopback
        or addr.is_link_local
        or addr.is_reserved
        or addr.is_multicast
    )

File: src/lib/url-security.ts
import { isIP } from 'node:net'

export interface UrlSecurityOptions {
  allowPrivateUrls: boolean
}

/**
 * Reject private, link-local, and metadata hosts unless explicitly allowed.
 * Blocks SSRF-style targeting of internal services from agent configs.
 */
export function validateRemoteUrl(url: URL, options: UrlSecurityOptions): void {
  if (options.allowPrivateUrls) return

  const host = normalizeHostname(url.hostname)
  if (isRestrictedHost(host)) {
    throw new Error(
      `Refusing to connect to private or restricted host "${host}"; ` +
        'pass --allow-private-urls for local/dev targets.',
    )
  }
}

function normalizeHostname(hostname: string): string {
  const trimmed = hostname.trim().toLowerCase()
  if (trimmed.startsWith('[') && trimmed.endsWith(']')) {
    return trimmed.slice(1, -1)
  }
  return trimmed
}

function isRestrictedHost(host: string): boolean {
  if (host === 'localhost' || host.endsWith('.localhost')) return true
  if (host === '0.0.0.0' || host === '::') return true

  const version = isIP(host)
  if (version === 4) return isPrivateIpv4(host)
  if (version === 6) return isPrivateIpv6(host)

  return false
}

function isPrivateIpv4(host: string): boolean {
  const parts = host.split('.').map((part) => Number(part))
  if (parts.some((part) => Number.isNaN(part) || part < 0 || part > 255)) return false

  const [a, b] = parts
  if (a === 10) return true
  if (a === 127) return true
  if (a === 169 && b === 254) return true
  if (a === 172 && b >= 16 && b <= 31) return true
  if (a === 192 && b === 168) return true
  if (a === 100 && b >= 64 && b <= 127) return true
  if (a === 0) return true
  return false
}

function isPrivateIpv6(host: string): boolean {
  const normalized = host.toLowerCase()
  if (normalized === '::1') return true
  if (normalized.startsWith('fc') || normalized.startsWith('fd')) return true
  if (normalized.startsWith('fe80')) return true
  return false
}
